rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -0,0 +1,490 @@
+* musl libc Release Builds
+* Channel Implementation Regresses Topology Throughput
+* Lua Transform Leaked Memory
+
+#### Upsides
+
+* We continue to reap the benefits of our criterion work.
+* We do not have to substantially change our approach to performance work.
+
+#### Downsides
+
+* Without improving the reliability of test-harness data we will continue to
+  find its results difficult to act on.
+* If we do not improve Vector's self-diagnostic capability we will struggle to
+  understand user's on-prem issues, currently a very high-touch process. As our
+  user base expands this problem will become more acute.
+* As our criterion tests increase in coverage the build time will balloon. This
+  will steadily drain our productivity as iteration loop time increases.
+
+### Lean Further Into Criterion
+
+In this alternative we make no substantial changes to the test-harness -- follow
+"do nothing" here -- but place more emphasis on the criterion work. In
+particular, we intend:
+
+* to increase the amount of compute available to the criterion CI, throwing
+  more hardware at the CI time issue,
+* to build benchmarks that demonstrate key components' throughput performance,
+  ensuring that these numbers are maintained in documentation for end users,
+* to substantially improve the coverage of benchmarked Vector code, though as
+  with correctness tests exact thresholds are a matter for team debate,
+* to run our criterion benchmarks across all supported Vector platforms and
+* to explore alternatives to improve criterion to derive better, more stable
+  signals on from our benchmarks.
+
+Addresses:
+
+* Regression from Increased Instrumentation
+* `metrics` Crate Upgrade Regresses Benchmarks
+
+Does not address:
+
+* musl libc Release Builds
+* Channel Implementation Regresses Topology Throughput
+* Lua Transform Leaked Memory
+
+#### Upsides
+
+* We reap the benefits of broader adoption of criterion in our project, which
+  include catching some regressions, offering targeted feedback to engineers
+  (if a test is, itself, targeted) and improve the broader ecosystem by
+  rolling changes into criterion.
+* We gain a good deal of detailed insight into Vector at a unit level.
+* We gain documented performance expectations for major components, a boon for
+  our end users when evaluating Vector.
+
+#### Downsides
+
+* As we have seen in practice, micro-benchmarks may not be representative of
+  macro-performance.
+* Wall-clock benchmarks are extremely sensitive to external factors.
+* We will encounter situations where benchmarks improve in CI and regress on
+  user's machines, especially as benchmarks become more "micro".
+
+### Build a `vector diagnostic` sub-command
+
+In this alternative we extend the Vector interface to include a `diagnostic`
+sub-command. This diagnostic will examine the system to gather information about
+it's running environment and perform, time fundamental actions. Information we
+might want to collect:
```

Comment:
Fully agreed, very complementary. I guess it's worth noting that my working style is very much to address problems as they pop up, addressing a class of problems when you can and that kind of work is what I was imagining here. I would like, eventually, to have vector be able to tell us about a baseline of capability in its deploy environment and I'd also like for vector to be able to answer questions about its interior life. Whichever seems to us to be of the most utility and the most feasible to get built strikes me as the one to pursue sooner. Either would be a boon.

Reply:
Yep, all of that makes sense. My bias towards measuring where time is being spent is just that it's a little more helpful on its own. It could be quite difficult to replicate an environment matching a given set of measurements, while "the file source is spending most of its time writing checkpoints" helps us narrow the problem quickly. Again, it's really just prioritization and the more we can collect the better.